src/FffGcodeWriter.cpp

```diff
@@ -1660,48 +1732,70 @@ bool FffGcodeWriter::addSupportInfillToGCode(const SliceDataStorage& storage, La
     EFillMethod support_pattern = infill_extr.getSettingAsFillMethod("support_pattern"); // first layer pattern must be same as other layers
     if (layer_nr <= 0 && (support_pattern == EFillMethod::LINES || support_pattern == EFillMethod::ZIG_ZAG)) { support_pattern = EFillMethod::GRID; }
 
-    int infill_extruder_nr_here = (layer_nr <= 0)? getSettingAsIndex("support_extruder_nr_layer_0") : getSettingAsIndex("support_infill_extruder_nr");
+    int infill_extruder_nr_here = (layer_nr <= 0) ? getSettingAsIndex("support_extruder_nr_layer_0") : getSettingAsIndex("support_infill_extruder_nr");
     const ExtruderTrain& infill_extr_here = *storage.meshgroup->getExtruderTrain(infill_extruder_nr_here);
 
-    const Polygons& support = support_layer.supportAreas;
-
-    std::vector<PolygonsPart> support_islands = support.splitIntoParts();
-
+    // create a list of outlines and use PathOrderOptimizer to optimize the travel move
+    std::vector<Polygons> infill_outline_list;
+    for (unsigned int part_idx = 0; part_idx < support_layer.support_infill_parts.size(); ++part_idx)
+    {
+        infill_outline_list.push_back(support_layer.support_infill_parts[part_idx].outline);
+    }
     PathOrderOptimizer island_order_optimizer(gcode_layer.getLastPosition());
-    for(unsigned int n=0; n<support_islands.size(); n++)
+    for (unsigned int wall_idx = 0; wall_idx < infill_outline_list.size(); ++wall_idx)
     {
-        island_order_optimizer.addPolygon(support_islands[n][0]);
+        island_order_optimizer.addPolygon(infill_outline_list[wall_idx][0]);
     }
     island_order_optimizer.optimize();
 
-    for(unsigned int n=0; n<support_islands.size(); n++)
+    // process each support infill islands
+    for (int support_infill_part_idx : island_order_optimizer.polyOrder)
     {
-        PolygonsPart& island = support_islands[island_order_optimizer.polyOrder[n]];
+        const SupportInfillPart& support_infill_part = support_layer.support_infill_parts[support_infill_part_idx];
 
-        int support_infill_overlap = 0; // support infill should not be expanded outward
-        
-        int offset_from_outline = 0;
+        // add outline (boundary) if the infill pattern is not Zig-Zag
+        const Polygons& outline = support_infill_part.outline;
         if (support_pattern == EFillMethod::GRID || support_pattern == EFillMethod::TRIANGLES || support_pattern == EFillMethod::CONCENTRIC)
         {
-            Polygons boundary = island.offset(-support_line_width / 2);
+            Polygons boundary = outline.offset(-support_line_width / 2);
```

Comment:
What are you doing here?

I think you should just add the outer perimeter.
Just add `support_infill_part.insets[0]`.
Or better yet: add all the insets, rather than only the outer one.